src/trace.rs

```diff
@@ -0,0 +1,212 @@
+// For completeness, wrappers around all of tracing's public macros are provided, even if they are
+// not used at the present time.
+#![allow(unused_macros)]
+
+//! Internal Tracing macro module
+//!
+//! The [`trace`][crate::trace] module is an internal module that contains wrapper macros encapsulating
+//! [`tracing`][tracing]'s macros. These macros allow for conditional expansion of
+//! [`tracing`][tracing] macros during compilation when the `tracing` feature is enabled, or trimming
+//! them when the feature is disabled, all in a concise manner.
+//!
+//! The macros from the [`trace`][crate::trace] module are declared by default and can be used throughout the
+//! crate. However, as the contents of these macros are conditionally compiled, they are effectively trimmed
+//! during inline expansion when the `tracing` feature is disabled.
+//!
+//! # Unstable
+//!
+//! The [`tracing`][tracing] module is currenty **unstable**, hence the existence of this module. As a
+//! result, hyper's [`tracing`][tracing] logs are only accessibe if `--cfg hyper_unstable_tracing` is
+//! passed to `rustc` when compiling. The easiest way to do that is through setting the `RUSTFLAGS`
+//! enviornment variable.
+//!
+//! # Building
+//!  
+//! Enabling [`trace`][crate::trace] logs, can be done with the following `cargo` command, as of
+//! version `1.64.0`:
+//!
+//! ```notrust
+//! RUSTFLAGS="--cfg hyper_unstable_tracing" cargo rustc --features client,http1,http2,tracing --crate-type cdylib
+//! ```
+
+macro_rules! debug {
+    ($($arg:tt)+) => {
+        #[cfg(feature = "tracing")]
+        {
+            #[cfg(not(hyper_unstable_tracing))]
+            compile_error!(
+                "\
+                The `tracing` feature is unstable, and requires the \
+                `RUSTFLAGS='--cfg hyper_unstable_tracing'` environment variable to be set.\
+            "
+            );
+            tracing::debug!($($arg)+);
+        }
+    }
+}
+
+macro_rules! debug_span {
+    ($($arg:tt)*) => {
+        #[cfg(feature = "tracing")]
+        {
+            #[cfg(not(hyper_unstable_tracing))]
+            compile_error!(
+                "\
+                The `tracing` feature is unstable, and requires the \
+                `RUSTFLAGS='--cfg hyper_unstable_tracing'` environment variable to be set.\
+            "
+            );
+            let span = tracing::debug_span!($($arg)+);
+            let _ = span.enter();
```

Comment:
I believe this will drop the enter guard immediately, since the block ends on the next line. You could change this to return `span.entered()`, and then the usage would be `let _s = debug_span!(blah);`.

Reply:
yeah, that's correct. we should either have the macro return an un-entered span that's entered at the callsite, or change this to 
```suggestion
            let _enter = span.enter();
```
or similar.

Reply:
The issue with `let _s = debug_span!(blah);` is that the compiler does not support removing an expression. When tracing is off, this effectively becomes `let _s = ;`. 

Similarly, `let _enter  = span.enter()` is also dropped at the end of the block.

The following seems to work:
```rust
use tracing_subscriber::prelude::*;

macro_rules! test {
    ($($arg:tt)+) => {
        #[cfg(debug_assertions)]
        { 
            let _sp = tracing::error_span!($($arg)+);
            _sp.entered()
        }
    };
}

fn main() {
    let stdout = tracing_subscriber::fmt::layer().pretty();

    tracing_subscriber::registry()
        .with(
            stdout
                .with_span_events(tracing_subscriber::fmt::format::FmtSpan::EXIT)
        )
        .init();
     
    let _e = { test!("asd", value = 32 ) };
    
    tracing::error!("test one");
    tracing::error!("test two");
}
```

Should I move forward with this approach?